### ISSUE TITLE
Attached enum emulating code with set and Symbol crashes online and using Linux command line

### ISSUE THREAD

**Issue by David-Else** [google/closure-compiler#3058]
```
Version: v20180805
Built on: 2018-08-09 20:09
```
switch to advanced and compile:

https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%2540language_out%2520ES5%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F%252F%2520ADD%2520YOUR%2520CODE%2520HERE%250Aconst%2520nameSpace%2520%253D%2520%257B%250A%2520%2520%252F**%250A%2520%2520%2520*%2520%2540enum%2520%257BSymbol%257D%250A%2520%2520%2520*%252F%250A%2520%2520SCREEN%253A%2520%257B%250A%2520%2520%2520%2520TITLE_SCREEN%253A%2520Symbol('TITLE_SCREEN')%252C%250A%2520%2520%2520%2520GAME_SCREEN%253A%2520Symbol('GAME_SCREEN')%252C%250A%2520%2520%2520%2520PAUSE%253A%2520Symbol('PAUSE')%252C%250A%2520%2520%2520%2520GAMEOVER_SCREEN%253A%2520Symbol('GAMEOVER_SCREEN')%252C%250A%2520%2520%257D%252C%250A%250A%2520%2520set%2520setScreen(key)%2520%257B%250A%2520%2520%2520%2520if%2520(key%2520in%2520this.SCREEN)%2520%257B%250A%2520%2520%2520%2520%2520%2520this.currentScreen%2520%253D%2520this.SCREEN%255Bkey%255D%253B%250A%2520%2520%2520%2520%257D%2520else%2520%257B%250A%2520%2520%2520%2520%2520%2520throw%2520new%2520TypeError(%2560%2524%257Bkey%257D%2520is%2520invalid%2520key%2520for%2520SCREEN%2520object%2560)%253B%250A%2520%2520%2520%2520%257D%250A%2520%2520%257D%252C%250A%257D%253B%250A%250AnameSpace.setScreen%2520%253D%2520'TITLE_SCREEN'%253B%250Aconsole.log(nameSpace.currentScreen)%253B%250A%250A%250A

```
const nameSpace = {
  /**
   * @enum {Symbol}
   */
  SCREEN: {
    TITLE_SCREEN: Symbol('TITLE_SCREEN'),
    GAME_SCREEN: Symbol('GAME_SCREEN'),
    PAUSE: Symbol('PAUSE'),
    GAMEOVER_SCREEN: Symbol('GAMEOVER_SCREEN'),
  },

  set setScreen(key) {
    if (key in this.SCREEN) {
      this.currentScreen = this.SCREEN[key];
    } else {
      throw new TypeError(`${key} is invalid key for SCREEN object`);
    }
  },
};

nameSpace.setScreen = 'TITLE_SCREEN';
console.log(nameSpace.currentScreen);
```
It crashes with and without the JSDoc part, but with the JSDoc it says:

```
compiler --compilation_level ADVANCED --language_out ES5 ./closureTest.js
./closureTest.js:6: WARNING - assignment to property TITLE_SCREEN of enum{nameSpace.SCREEN}
found   : symbol
required: Symbol
    TITLE_SCREEN: Symbol('TITLE_SCREEN'),
    ^^^^^^^^^^^^

./closureTest.js:7: WARNING - assignment to property GAME_SCREEN of enum{nameSpace.SCREEN}
found   : symbol
required: Symbol
    GAME_SCREEN: Symbol('GAME_SCREEN'),
    ^^^^^^^^^^^

./closureTest.js:8: WARNING - assignment to property PAUSE of enum{nameSpace.SCREEN}
found   : symbol
required: Symbol
    PAUSE: Symbol('PAUSE'),
    ^^^^^

./closureTest.js:9: WARNING - assignment to property GAMEOVER_SCREEN of enum{nameSpace.SCREEN}
found   : symbol
required: Symbol
    GAMEOVER_SCREEN: Symbol('GAMEOVER_SCREEN'),
    ^^^^^^^^^^^^^^^
```

**Comment by lauraharker** [comment]
Thanks for the report.

The warnings with the Symbol JSDoc are because `/** @enum {Symbol} */` should really be `/** @enum {symbol} */`.

`symbol` is the primitive type you get by calling `Symbol('TITLE_SCREEN')`. `Symbol` really shouldn't be a valid type IMO, since `Symbol` is not a constructor, but an artifact of how the Closure Compiler represents types internally means that you can use `Symbol` as a type.

The crash seems to be a bug with setters in object literals. [This also crashes](https://closure-compiler-debugger.appspot.com/#input0%3Dconst%2520nameSpace%2520%253D%2520%257B%250A%250A%2520%2520set%2520setScreen(key)%2520%257B%250A%2520%2520%2520%2520alert('setting%2520the%2520screen')%253B%250A%2520%2520%257D%252C%250A%257D%253B%250A%250AnameSpace.setScreen%2520%253D%2520'TITLE_SCREEN'%253B%26input1%26conformanceConfig%26externs%3Dfunction%2520alert(s)%2520%257B%257D%26refasterjs-template%26includeDefaultExterns%3Dtrue%26CHECK_TYPES%3Dtrue%26TRANSPILE%3Dtrue%26COLLAPSE_PROPERTIES%3Dtrue%26CLOSURE_PASS%3Dtrue%26PRESERVE_TYPE_ANNOTATIONS%3Dtrue%26PRETTY_PRINT%3Dtrue):
``` javascript
const nameSpace = {
  set setScreen(key) {
    alert('setting the screen');
  },
};
nameSpace.setScreen = 'TITLE_SCREEN';
```

**Comment by David-Else** [comment]
Thanks so much! It compiles!

Unfortunately I have a new problem now, but I don't know if it is related. The code compiles in advance mode to:
```
    h: {
      M: Symbol('TITLE_SCREEN'),
      B: Symbol('GAME_SCREEN'),
      ga: Symbol('PAUSE'),
      U: Symbol('GAMEOVER_SCREEN'),
    },
    set l(a) {
      if (a in this.h) this.i = this.h[a];
      else throw new TypeError(`${a} is invalid key for SCREEN object`);
    },
```
but I get an error:
```
bundle-compressed.js:266 Uncaught TypeError: TITLE_SCREEN is invalid key for SCREEN object
    at Object.set l [as l] (bundle-compressed.js:266)
```
Which comes from:

`b.l = 'TITLE_SCREEN';`

Which is no good as this string that I was using to refer to the key has been renamed to 'M' by the complier. What can I do to prevent this happening? Thanks!

**Comment by lauraharker** [comment]
In ADVANCED mode, the compiler assumes that properties aren't referred to with a mix of dotted names and quoted names/computed properties. It will rename all non-quoted properties but won't try to change string literal references to those properties.

The easiest solution in your case would be to switch `TITLE_SCREEN: Symbol('TITLE_SCREEN')` to `'TITLE_SCREEN: Symbol('TITLE_SCREEN')`.

Additional information on property renaming/compiler assumptions:
https://developers.google.com/closure/compiler/docs/api-tutorial3#propnames
http://closuretools.blogspot.com/2012/09/which-compilation-level-is-right-for-me.html
https://developers.google.com/closure/compiler/docs/limitations

**Comment by ChadKillingsworth** [comment]
Your setter also uses the keyword `this`. That's not generally safe in Advanced mode outside of classes. Use `namespace` directly.

**Comment by David-Else** [comment]
@ChadKillingsworth thanks for the extra input! I changed the whole object into a class to avoid any potential problems with `.this`.

I made the changes @lauraharker suggested to use quotes in the object keys, and it did compile but the program did not work properly. I re-factored out the quoted names/computed properties into dot notation and that did the job.

This is my first attempt at using closure compiler, and now I can see that with a bit of work it is the ultimate minifier. I  am learning in honour of the http://js13kgames.com/ competition that is running. Thanks again for the help!
